src/components/form-elements/draft-js-mention-selector/DraftJSMentionSelector.js

```diff
@@ -75,27 +77,83 @@ class DraftJSMentionSelector extends React.Component<Props, State> {
-     * @returns {void}
-     */
-    componentWillReceiveProps(nextProps: Props) {
+    static getDerivedStateFromProps(nextProps: Props) {
         const { contacts } = nextProps;
+        return contacts ? { contacts } : null;
+    }
+
+    componentDidUpdate(prevProps: Props, prevState: State) {
+        const { internalEditorState: prevInternalEditorState } = prevState;
+        const { internalEditorState } = this.state;
+
+        const { editorState: prevEditorStateFromProps } = prevProps;
+        const { editorState } = this.props;
+
+        // Determine whether we're working with the internal editor state or
+        // external editor state passed in from props
+        const prevEditorState = prevInternalEditorState || prevEditorStateFromProps;
+        const currentEditorState = internalEditorState || editorState;
 
-        if (contacts) {
-            this.setState({ contacts }, () => {
-                this.checkValidity();
-            });
+        // Only handle isTouched state transitions and check validity if the
+        // editorState references are different. This is to avoid getting stuck
+        // in an infinite loop of checking validity because checkValidity always
+        // calls setState({ error })
+        if (prevEditorState && currentEditorState && prevEditorState !== currentEditorState) {
+            const newState = this.getDerivedStateFromEditorState(currentEditorState, prevEditorState);
+            if (newState) {
+                this.setState(newState, this.checkValidityIfAllowed);
+            } else {
+                this.checkValidityIfAllowed();
+            }
         }
     }
 
+    getDerivedStateFromEditorState(currentEditorState: EditorState, previousEditorState: EditorState) {
+        const isPreviousEditorStateEmpty = this.isEditorStateEmpty(previousEditorState);
+        const isCurrentEditorStateEmpty = this.isEditorStateEmpty(currentEditorState);
+        const isNewEditorState = isCurrentEditorStateEmpty && !isPreviousEditorStateEmpty;
+        const isEditorStateDirty = isPreviousEditorStateEmpty && !isCurrentEditorStateEmpty;
+
+        let newState = null;
+        // Detect case where controlled EditorState is created anew and empty.
+        // If next editorState is empty and the current editorState is not empty
+        // that means it is a new empty state and this component should not be marked dirty
+        if (isNewEditorState) {
+            newState = { isTouched: false, error: null };
+        } else if (isEditorStateDirty) {
+            // Detect case where controlled EditorState has been made dirty
+            // If the current editorState is empty and the next editorState is not
+            // empty then this is the first interaction so mark this component dirty
+            newState = { isTouched: true };
+        }
+
+        return newState;
+    }
+
+    checkValidityIfAllowed() {
+        const { validateOnBlur }: Props = this.props;
```

Comment:
can default it to noop to reduce a branch

Reply:
In this case, I want the branch because I only want to `checkValidity` if `validateOnBlur` if false. If it's true, then on through `handleBlur` will `checkValidity` be called